server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,229 @@
+import io.reactivex.rxjava3.core.Flowable;
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.query.remote.ProtobufMetadataManager;
+import org.infinispan.query.remote.client.ProtobufMetadataManagerConstants;
+import org.infinispan.query.remote.json.JsonQueryErrorResult;
+import org.infinispan.rest.InvocationHelper;
+import org.infinispan.rest.NettyRestResponse;
+import org.infinispan.rest.cachemanager.RestCacheManager;
+import org.infinispan.rest.framework.ContentSource;
+import org.infinispan.rest.framework.ResourceHandler;
+import org.infinispan.rest.framework.RestRequest;
+import org.infinispan.rest.framework.RestResponse;
+import org.infinispan.rest.framework.impl.Invocations;
+import org.infinispan.rest.operations.exceptions.NoDataFoundException;
+import org.infinispan.rest.operations.exceptions.NoKeyException;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_JSON;
+import static org.infinispan.rest.framework.Method.DELETE;
+import static org.infinispan.rest.framework.Method.GET;
+import static org.infinispan.rest.framework.Method.POST;
+import static org.infinispan.rest.framework.Method.PUT;
+
+/**
+ * Protobuf schema manipulation Resource
+ *
+ * @author Katia Aresti
+ * @since 11
+ */
+public class ProtobufResource extends BaseCacheResource implements ResourceHandler {
+
+   private final ObjectMapper objectMapper;
+
+   public ProtobufResource(InvocationHelper invocationHelper) {
+      super(invocationHelper);
+      objectMapper = invocationHelper.getMapper();
+   }
+
+   @Override
+   public Invocations getInvocations() {
+      return new Invocations.Builder()
+            // Key related operations
+            .invocation().methods(GET).path("/v2/schema").handleWith(this::getSchemasNames)
+            .invocation().methods(POST).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,true))
+            .invocation().methods(PUT).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,false))
+            .invocation().methods(GET).path("/v2/schema/{schemaName}").handleWith(this::getSchema)
+            .invocation().method(DELETE).path("/v2/schema/{schemaName}").handleWith(this::deleteSchema)
+            .create();
+   }
+
+   private CompletionStage<RestResponse> getSchemasNames(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      return Flowable.fromIterable(cache.entrySet())
```

Comment:
@wburns this is not ok ?

Reply:
No, unfortunately the iterator returned from `cache.entrySet()` is blocking. This is why normally the `ClusterPublisher` is used as it allows for non blocking "iteration". But unfortunately without the new API, there is no easy way to do it with media type conversions.

I think the simplest way for now is just to delegate this call to the blocking thread pool.

Reply:
That is unless the PROTOBUF_METADATA_CACHE_NAME doesn't do any media type conversions, then you can just use the ClusterPublisherManager directly.

Reply:
ClusterPublisherManager does not work with EncoderCache instances?

Reply:
It is not exposed via an EncoderCache in any way. It is solely a component currently.

Reply:
The problem is I was expecting to effectively expose this from the new API, so I didn't add anything for out internal use. Maybe we should add an internal method only somewhere. The issue with ClusterPublisherManager is that the API of it is probably unwieldy to power users and we normally hide a lot of that from the user.

Reply:
So this one should be changed as well?
https://github.com/infinispan/infinispan/blob/1fc5a2ef2a4ae2876ee52a05b757da9b322641e0/server/rest/src/main/java/org/infinispan/rest/resources/CacheManagerResource.java#L221

Reply:
> So this one should be changed as well?
> https://github.com/infinispan/infinispan/blob/1fc5a2ef2a4ae2876ee52a05b757da9b322641e0/server/rest/src/main/java/org/infinispan/rest/resources/CacheManagerResource.java#L221

I have to check the `cachesHealth` configuration. If it can be distributed or have a store, then yes it needs to be changed.

Reply:
Oh actually `cachesHealth` is just a HashMap :) However, https://github.com/infinispan/infinispan/blob/1fc5a2ef2a4ae2876ee52a05b757da9b322641e0/server/rest/src/main/java/org/infinispan/rest/resources/CacheManagerResource.java#L214 can block as mentioned in the comment.

Reply:
As I mentioned if we add blockhound to server/rest we won't have to conjecture about these and it will just fail the test if it blocks :)

Reply:
Check out https://github.com/infinispan/infinispan/pull/8323